Add test coverage to myjsonrpc

Wabri opened the pull request:
Implementing missing tests for myjsonrpc:

- [myjsonrpc:43-44](https://app.codecov.io/gh/os-autoinst/os-autoinst/blob/master/myjsonrpc.pm#L43) => done
- [myjsonrpc:104-105](https://app.codecov.io/gh/os-autoinst/os-autoinst/blob/master/myjsonrpc.pm#L104) => set as uncover
- [myjsonrpc:111-112](https://app.codecov.io/gh/os-autoinst/os-autoinst/blob/master/myjsonrpc.pm#L111) => done

Issue: https://progress.opensuse.org/issues/167935

perlpunk commented on t/24-myjsonrpc.t:
```diff
@@ -3,6 +3,9 @@
 
+BEGIN {
+    $ENV{PERL_MYJSONRPC_DEBUG} = 0;
+}
```
We do have `t/24-myjsonrpc-debug.t` for that already.

Wabri replied:
I set to 0 not to 1

perlpunk commented on myjsonrpc.pm:
```diff
@@ -38,10 +38,10 @@ sub send_json ($to_fd, $cmd) {
-    if (!$wb || $wb != length($json)) {
-        die('myjsonrpc: remote end terminated connection, stopping') if !DEBUG_JSON && $! =~ qr/Broken pipe/;
-        confess "syswrite failed: $!";
+    my $wb = syswrite($to_fd, "$json"); # wb = write buffer
+    if (!$wb || $wb != length($json)) { # is true when wb is empty or when the lenght of wb is not the entire json lenght
```
Rather than commenting stuff we should aim for making the code clearer, for example by using descriptive variable names.
In this case it seemed to cause a misunderstanding. `$wb` is not the buffer, but the return value of `syswrite`, which is `Returns the number of bytes actually written`.
With that I hope we can go without a comment. `!$number` is basically short for `$number == 0` in perl:
```suggestion
    my $written_bytes = syswrite($to_fd, "$json");
    if (!$written_bytes || $written_bytes != length($json)) {
```

Wabri replied:
As a draft this is not a ready pr, I'll remove the comment while refactoring :+1: 

I thought that `!$wb` was used to store the output, thanks for the hint :pray: I'll rename the variable as you suggest :+1:

perlpunk commented on myjsonrpc.pm:
```diff
@@ -38,10 +38,10 @@ sub send_json ($to_fd, $cmd) {
-        confess "syswrite failed: $!";
+    my $wb = syswrite($to_fd, "$json"); # wb = write buffer
+    if (!$wb || $wb != length($json)) { # is true when wb is empty or when the lenght of wb is not the entire json lenght
+        die('myjsonrpc: remote end terminated connection, stopping') if !DEBUG_JSON && $! =~ qr/Broken pipe/; # terminate the script if we are not debugging and the syserror is Broken pipe 
+        confess "syswrite failed: $!"; # write to output the error
```
That's commenting obvious things.
Of course you might argue that it is not that obvious for a perl newbie, but then we would end up with thousands of comments in our code :)

Wabri replied:
Yes, I've never use perl so this helps me remember what I've read. As I mention above I'll remove every comments on my code that is not strictly necessary.

Martchus reviewed (commented):
The commit "[Update readability on send_json function](https://github.com/os-autoinst/os-autoinst/pull/2578/commits/70cb075540fdc561b106d543cdeee669d1115445) should probably be squashed with the commit that initially changed that function.

The rest looks good so far although https://github.com/os-autoinst/os-autoinst/blob/master/t/24-myjsonrpc-debug.t might be the better place to add those tests to.

perlpunk commented on t/24-myjsonrpc.t:
```diff
@@ -56,11 +59,28 @@ sub magic_close () {
     like($warnings[0], qr{received magic close});
 };
 
+subtest 'send_json dies when buffer is empty and pipe is broken' => sub {
+    local *CORE::GLOBAL::syswrite = sub {
+        $! = 'broken pipe';
+        return 0;
+    };
+    dies_ok { send_json($child, undef) };
+};
```
You can't easily override builtin functions in a different package like that.
But I think we can test this without overriding, by passing a filehandle that is not for writing:
```suggestion
subtest 'syswrite error handling' => sub {
    my @warnings = warnings {
        throws_ok {
            myjsonrpc::send_json(\*STDIN, {cmd => 'something'})
        } qr/syswrite failed: Bad file descriptor/, 'send_json dies when syswrite fails';
    };
    like $warnings[0], qr/Filehandle STDIN opened only for input/, 'got expected warning';
};

```

Wabri replied:
Miss that and yes, I've done several tests around mocking the syswrite buildin function, but with no results.

To solve it I've extracted from the send_json function the syswrite into another function in order to mock that and it works, I've updated the code right now!

Wabri replied:
As you seen in my latest commit:

```perl
subtest 'send_json dies when buffer is empty and pipe is broken' => sub {
    my $myjsonrpc_mock = Test::MockModule->new('myjsonrpc');
    $myjsonrpc_mock->redefine(get_written_bytes => sub ($to_fd, $json) {
            return 0;
    });
    dies_ok { myjsonrpc::send_json($child, $send1) } 'myjsonrpc: remote end terminated connection, stopping';
};
```

with get_written_bytes that is a function in myjsonrpc.pm that simply bind the syswrite buildin function:

```perl
sub get_written_bytes($to_fd, $json) {
    return syswrite($to_fd, "$json");
}
```

It is not elegant, what do you think?

perlpunk commented on t/24-myjsonrpc.t:
```diff
@@ -56,11 +59,28 @@ sub magic_close () {
+    dies_ok { send_json($child, undef) };
+};
+
+subtest 'send_json succeeds when buffer is empty and pipe is NOT broken' => sub {
```
The successful case is already tested, so we don't need that

Wabri replied:
Yes, I've already removed with my latest commits!

Wabri committed:
Add coverage test for written bytes errors

https://progress.opensuse.org/issues/167935

okurz commented on myjsonrpc.pm:
```diff
@@ -14,6 +14,10 @@ use bmwqemu ();
 use constant DEBUG_JSON => $ENV{PERL_MYJSONRPC_DEBUG} || 0;
 use constant READ_BUFFER => $ENV{PERL_MYJSONRPC_BYTES} || 8_000_000;
 
+sub get_written_bytes($to_fd, $json) {
+    return syswrite($to_fd, "$json");
+}
```
```suggestion
sub _syswrite($to_fd, $json) { syswrite($to_fd, "$json") }
```

Martchus replied:
And also avoid the string interpolation:

```suggestion
sub _syswrite($to_fd, $json) { syswrite($to_fd, $json) }
```

We could actually skip the argument processing here to make it even simpler and more lightweight:

```suggestion
sub _syswrite(@_) { syswrite(@_) }
```

Wabri committed:
Refactor utility functions for myjsonrpc

https://progress.opensuse.org/issues/167935

Co-authored-by: Oliver Kurz <okurz@suse.de>
Co-authored-by: Marius Kittler <mkittler@suse.de>

Wabri committed:
Set a bmwqemu diag uncoverable

We try a lot of solutions but by now every one of them bring to
complications or unreadable code.

https://progress.opensuse.org/issues/167935

Wabri committed:
Export handle_read_error function from read_json

https://progress.opensuse.org/issues/167935

Co-authored-by: Marius Kittler <mkittler@suse.de>

okurz commented:
what about the "not-ready" label?

okurz reviewed (changes requested):
Please also add the file myjsonrpc.pm in codecov in the "fully_covered" section

perlpunk commented on myjsonrpc.pm:
```diff
@@ -95,22 +98,11 @@ sub read_json ($socket, $cmd_token = undef, $multi = undef) {
-            bmwqemu::fctwarn("sysread failed: $!") if DEBUG_JSON || $bmwqemu::vars{DEBUG_JSON_RPC};
-            return;
-        }
+        if (!sysread($socket, $qbuffer, READ_BUFFER)) { bmwqemu::fctwarn("sysread failed: $!") if is_debug(), return }
```
You can use a semicolon here:
```suggestion
        if (!sysread($socket, $qbuffer, READ_BUFFER)) { bmwqemu::fctwarn("sysread failed: $!") if is_debug(); return }
```

perlpunk replied:
I still see a comma

Wabri replied:
Try now!

perlpunk reviewed (commented):
It's ok to keep commits separate if they are about different things, but I think in those 7 commits you rename and rearrange things multiple times, so I suggest to squash some commits.

perlpunk commented:
Also please rebase on current master. The coverage report shows several unrelated coverage changes.
It's a good idea to fetch origin and rebase a PR whenever you push new changes, daily.

Martchus commented on myjsonrpc.pm:
```diff
@@ -14,6 +14,21 @@ use bmwqemu ();
+
+sub is_debug () { DEBUG_JSON || $bmwqemu::vars{DEBUG_JSON_RPC} }
+
+sub wait_read ($fd) {
```
The function is not waiting for anything:
```suggestion
sub handle_read_error ($fd) {
```
(Of course the function `can_read` is also badly named because that one *does* the waiting and is not called `wait_…`.)

Martchus reviewed (commented):
Looks good, I only have a small nitpick on one of the function names.

Wabri commented:
Done rebase and solve all the code review

perlpunk commented on myjsonrpc.pm:
```diff
@@ -14,6 +14,21 @@ use bmwqemu ();
 use constant READ_BUFFER => $ENV{PERL_MYJSONRPC_BYTES} || 8_000_000;
 
+# hash for keeping state
+our $sockets;
```
This doesn't need to be a global package var

```suggestion
my $sockets;
```
* Fixed in #2960
